// playwright-gherkin/lib/DataTable.js
function transpose(matrix) {
    return matrix[0].map((col, i) => matrix.map(row => row[i]));
}
function objectify(matrix) {
    const keys = matrix[0];
    const values = matrix.slice(1);
    return values.map((values) => Object.fromEntries(keys.map((key, index) => [key, values[index]])));
}
export class DataTable {
    get colMajor() {
        if (this._colMajor)
            return this.colMajor;
        const colMajor = transpose(this.rowMajor);
        this._colMajor = colMajor;
        return colMajor;
    }
    get colObjects() {
        if (this._colObjects)
            return this.colObjects;
        const colObjects = objectify(this.colMajor);
        this._colObjects = colObjects;
        return colObjects;
    }
    get rowObjects() {
        if (this._rowObjects)
            return this.rowObjects;
        const rowObjects = objectify(this.rowMajor);
        this._rowObjects = rowObjects;
        return rowObjects;
    }
    constructor(rowMajor) {
        this.rowMajor = rowMajor;
    }
}
