// playwright-gherkin/lib/generate.js
import { SourceMapGenerator } from 'source-map';
const playwrightArgs = ['page', 'browser', 'context', 'request'];
function genStep(step) {
    return [
        { line: step.location?.line || 1, step: step.originalText },
        `await test.step(${JSON.stringify(step.originalText)}, async ()=>{`,
        [
            `const table = ${step.table ? `new DataTable(${JSON.stringify(step.table)})` : 'undefined'};`,
            `const docString = ${JSON.stringify(step.docString)};`,
            `const step = ${JSON.stringify(step)};`,
            `const {call, parameters} = steps.find(step);`,
            `const arg1 = {${playwrightArgs.join(', ')}, table, docString, step, world, parameters};`,
            `const timeout = info.timeout ? new Promise((_, rej)=>setTimeout(()=>rej(new Error('Step timeout reached after '+info.timeout+'ms')), info.timeout)) : undefined;`,
            `await Promise.race([call(arg1, info), timeout].filter(p=>p));`,
        ],
        `});`,
    ];
}
function genScenario(scn) {
    const tests = scn.steps.flatMap(step => genStep(step));
    return [
        { line: scn.location?.line || 1, step: scn.name },
        `test(${JSON.stringify(`${scn.name} ${scn.tags.join(' ')}`.trim())}, async ({${playwrightArgs.join(', ')}}, info)=>{`,
        [
            `test.setTimeout(${scn.steps.length} * info.timeout);`,
            `const world = JSON.parse(JSON.stringify(steps.defaultWorld ?? {}));`,
            ...tests,
        ],
        `})`,
    ];
}
function genFeature(spec) {
    const describes = spec.features[0].scenarios.map(scn => genScenario(scn));
    return [
        `/* DO NOT EDIT! THIS FILE WAS GENERATED BY 'playwright-gherkin'. */`,
        `import {test} from '@playwright/test';`,
        `import {DataTable} from 'playwright-gherkin';`,
        `import {steps} from './steps.js';`,
        //`validate(${JSON.stringify(uri)}, ${JSON.stringify(hash)})`,
        ``,
        { line: spec.features[0].location?.line || 1, step: spec.features[0].name },
        `test.describe(${JSON.stringify(`${spec.features[0].name} ${spec.features[0].tags.join(' ')}`.trim())}, ()=>{`,
        ...describes,
        `})`,
    ];
}
export function generateCode(spec) {
    const block = genFeature(spec);
    const { output, map } = flattenCodeBlock(block);
    const sourcemap = new SourceMapGenerator({
        file: spec.uri,
    });
    sourcemap.setSourceContent(spec.uri, spec.content);
    for (const { generated, original } of map) {
        sourcemap.addMapping({
            generated,
            original: { column: 0, ...original },
            source: spec.uri,
            name: original.step || undefined,
        });
    }
    // TODO check this code when https://github.com/microsoft/playwright/issues/21204 is closed or fixed
    return output.join('\n');
    //return output.join('\n')
    //  +'\n\n'
    //  +'//# sourceMappingURL=data:application/json;charset=utf-8;base64,'+btoa(sourcemap.toString());
}
function flattenCodeBlock(block, map = [], output = [], level = 0, currentLocation = { line: 1, step: '' }) {
    const indentation = new Array(level).fill(' ').join('');
    for (const line of block) {
        if (typeof line === 'string') {
            output.push(`${indentation}${line}`);
            map.push({
                original: currentLocation,
                generated: { line: output.length, column: 0 },
            });
        }
        if (typeof line === 'object') {
            if (line instanceof Array) {
                flattenCodeBlock(line, map, output, level + 4, currentLocation);
            }
            else {
                currentLocation = line;
            }
        }
    }
    return { map, output };
}
