// playwright-gherkin/lib/parse.js
import Gherkin from '@cucumber/gherkin';
import { IdGenerator } from '@cucumber/messages';
import { dialects } from '@cucumber/gherkin';
const uuidFn = IdGenerator.uuid();
const builder = new Gherkin.AstBuilder(uuidFn);
const matcher = new Gherkin.GherkinClassicTokenMatcher();
const parser = new Gherkin.Parser(builder, matcher);
const stepTypeToDialectKey = (type, dialect = dialects.en) => {
    const dialectKey = {
        'Context': 'given',
        'Action': 'when',
        'Outcome': 'then',
        'Conjunction': 'and',
        'Unknown': 'unknown',
    }[type];
    return dialect[dialectKey].find((k) => k !== '* ').trim();
};
export function parseFeature(uri, feature) {
    const document = parser.parse(feature);
    const pickles = Gherkin.compile(document, uri, uuidFn);
    const dialect = dialects[document.feature?.language ?? 'en'];
    return {
        uri,
        content: feature,
        comments: document.comments.map(c => c.text),
        language: document.feature?.language ?? 'en',
        features: document.feature ? [{
                language: document.feature.language ?? 'en',
                description: document.feature?.description,
                tags: document.feature.tags.map(t => t.name),
                keyword: document.feature.keyword,
                name: document.feature.name,
                location: document.feature.location,
                scenarios: pickles.map(pickle => {
                    const astNode = document.feature?.children
                        .map(child => child.scenario)
                        .find(scn => pickle.astNodeIds.includes(scn?.id ?? ''));
                    const scenariosWithSameName = pickles.filter(p => p.name === pickle.name);
                    const scenarioIndex = scenariosWithSameName.indexOf(pickle);
                    const outlineSuffix = scenariosWithSameName.length > 1 ? ` (Example ${scenarioIndex + 1})` : '';
                    return {
                        name: pickle.name + outlineSuffix,
                        tags: pickle.tags.map(t => t.name),
                        location: astNode?.location,
                        steps: pickle.steps.map(step => {
                            const astNode = document.feature?.children
                                .flatMap(c => [...c.background?.steps ?? [], ...c.scenario?.steps ?? []])
                                .find(s => step.astNodeIds.includes(s?.id ?? ''));
                            const originalKeyword = astNode?.keyword.trim() ?? '';
                            const keyword = stepTypeToDialectKey(step.type ?? 'Unknown', dialect);
                            const originalText = `${originalKeyword} ${astNode?.text.trim()}`;
                            const text = keyword + ' ' + step.text;
                            return {
                                location: astNode?.location,
                                type: step.type ?? 'Unknown',
                                text,
                                keyword,
                                originalKeyword,
                                originalText,
                                tokens: tokenize(text),
                                table: step.argument?.dataTable?.rows.map(row => row.cells.map(cell => cell.value)),
                                docString: step.argument?.docString?.content,
                            };
                        })
                    };
                })
            }] : []
    };
}
export function tokenize(text) {
    const [U, _Q1, V, _Q2, ...X] = text.split(/( "|" |"$)/g);
    return [
        ...U.split(' '),
        ...(V ? [V] : []),
        ...(X.filter(x => x).length ? tokenize(X.join('')) : [])
    ];
}
export function parseStep(text, dialect) {
    const andKeyword = dialect.and.find(keyword => text.startsWith(keyword))?.trim();
    if (andKeyword)
        return { type: 'Conjunction', keyword: andKeyword, text, tokens: tokenize(text) };
    const givenKeyword = dialect.given.find(keyword => text.startsWith(keyword))?.trim();
    if (givenKeyword)
        return { type: 'Context', keyword: givenKeyword, text, tokens: tokenize(text) };
    const whenKeyword = dialect.when.find(keyword => text.startsWith(keyword))?.trim();
    if (whenKeyword)
        return { type: 'Action', keyword: whenKeyword, text, tokens: tokenize(text) };
    const thenKeyword = dialect.then.find(keyword => text.startsWith(keyword))?.trim();
    if (thenKeyword)
        return { type: 'Outcome', keyword: thenKeyword, text, tokens: tokenize(text) };
    throw new Error('Unable to parse: ' + text);
}
