// playwright-gherkin/lib/StepRegistry.js
import { dialects } from '@cucumber/gherkin';
import { parseStep } from './parse.js';
export class StepRegistry {
    constructor(dialect = process.env.npm_package_config_gherkin_dialect || 'en', defaultWorld) {
        this.defaultWorld = defaultWorld;
        this.steps = new Map();
        if (!(dialect in dialects))
            throw new Error('Invalid Dialect');
        this.dialect = dialects[dialect];
    }
    find({ tokens, text }) {
        let definedSteps = [...this.steps.entries()].filter(([definedTokens]) => definedTokens.length === tokens.length);
        const parameters = [];
        for (const [index, token] of tokens.entries()) {
            definedSteps = definedSteps.filter(([definedTokens]) => {
                if (definedTokens[index] === '{}') {
                    const parameterValue = token.startsWith('"') && token.endsWith('"') ? token.slice(1, -1) : token;
                    parameters.push(parameterValue);
                    return true;
                }
                return definedTokens[index].split('/').includes(token);
            });
        }
        if (definedSteps.length === 1)
            return {
                call: definedSteps[0][1],
                tokens,
                parameters
            };
        if (definedSteps.length >= 1)
            throw new Error(`Found multiple steps for: ${text}`);
        throw new Error(`Unable to find step: ${text}`);
    }
    ;
    define(statement, step) {
        const parsed = parseStep(statement, this.dialect);
        if (parsed.type === 'Conjunction')
            throw new Error('Cannot start a step with a conjunction');
        try {
            this.find(parsed);
        }
        catch {
            return this.steps.set(parsed.tokens, step);
        }
        throw new Error('Step already exists');
    }
    ;
}
